Enhance: add per profile dictionary capability

SlySven opened the pull request:
It is saved at the end of a session and the affix file (`profile.aff`) is maintained to allow for the Hunspell suggestion capability to operate. The word-list (`profile.dic`) file is checked at session start to accommodate manual editing (adding/removing words) between sessions.

When a word typed on the command line is not found in the main (system provided on non-Windows OSes) dictionary, instead of the red (now wavy on Unix, dotted on macOS) underline a cyan dashed underline is used to show that the word was found in the profile's own dictionary. Words can also be added and removed.

There are added options to populate the profile dictionary with all words sent by the MUD Server that do not appear in the selected system dictionary, with options to require a minimum character (actually *grapheme*) count in a word before it is so added and also an option to not add words containing digits. Other options may be considered later.

Also found out why the wavy / dashed underline was not showing in previous/initial attempt.

Some lua commands are also added:
* `addWordToProfileDictionary((string) word)` returns true if the word was added (and was not already in the dictionary)
* `removeWordFromProfileDictionary((string) word)` returns true if the word was present and removed from the dictionary.
* `getProfileDictionaryWordList()` returns a (sorted into system locale case insensitive order) list of ALL words in the profile dictionary.
* `spellCheckWord((string) word [,(bool)useProfileDictionary])` do a check in the specified system dictionary (default) or if a second argument is supplied and is `true` use the per profile one. Returns `true` if the word is found in the dictionary, and false if not.
* `spellSuggestWord((string) word [,(bool)useProfileDictionary])` does the same sort of suggestions search as is done on the console commmand line and returns a list of suggestions from the specified system dictionary (default) or if a second argument is supplied and is `true` use the per profile one.

**ALL** spell-checking activity is ceased if the option is disabled in the preferences (this was not the case previously)!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
> a cyan dashed underline is used to show that the word was found in the profile's own dictionary

From a UI point of view - that is not a necessary feature to have, in fact I think it'll hurt more than help because people will be confused. My take on it is to stick with the true and tried paradigm that works well: only highlight if the word is not found in _any_ dictionaries. Do others from @Mudlet/ui-review have thoughts on this?

Kebap commented on src/ui/profile_preferences.ui:
```diff
@@ -568,27 +568,109 @@
+                <item alignment="Qt::AlignLeft">
+                 <widget class="QLabel" name="label_autoAddDictionaryWordMinLength_before">
+                  <property name="text">
+                   <string extracomment="This (QLabel) texts is placed before a QSpinBox that displays a number with another QLabel that finishes a sentence">Minimum length</string>
```
Please don't interrupt sentences like this
```
Minimum length
4
characters
```

How about 
```
Minimum length:
4
```

Or minimum length in characters. 
Or minimum length of words to add.
Yeah I like the last one best.

SlySven replied:
Yeah, it is not a sane idea :stuck_out_tongue_winking_eye: to split the text like this - it might look pretty in English but it is harder to localize...

... I ***will*** fix this!

SlySven replied:
This part has been nuked anyhow so this is no longer an issue.

Kebap commented on src/ui/profile_preferences.ui:
```diff
@@ -568,27 +568,109 @@
              <bool>true</bool>
             </property>
+            <property name="checked">
+             <bool>false</bool>
```
Before this PR, default was spell check enabled. Did we loose this now?

SlySven replied:
That is a bad idea (IMVHO) because not all languages *have* spell checking support - I have come to realise that some of the ideographic scripts (think Far-Eastern!) cannot currently be handled by Hunspell.  As it is, I am just running through the list of languages that my Linux distribution supports to see whether it would be reasonable to grab them to fix #2291 ...

SlySven replied:
It turns out it is only like that if there is not a `Host` instance around - to leave the display blank and disabled - it gets set to the current state of the relevant `Host` when that is used to initialise the preferences....

Kebap commented:
I am with Vadi here: 
> only highlight if the word is not found in any dictionaries.

Kebap commented:
Explain to me again why this should be profile specific? If I have multiple profiles connected to a same game, they all need to learn their words seperately? Not sure how that's benefiting?!

Kebap commented:
> spellCheckWord((string) word [,(bool)useProfileDictionary]) do a check in the specified system dictionary (default) 

How does this default function on OS like Windows where there is no system dictionary provided?

vadi2 commented:
We provide them on Windows (see the hunspell.aff and hunspell.dic).

vadi2 commented:
I can't reply to the comment in-line for some reason - but spellcheck definitely should be **on** by default - at most, on by default for languages at have a dictionary for. Turning it off by default for English will be a regression.

vadi2 commented:
> populate the profile dictionary with all words sent by the MUD Server that do not appear in the selected system dictionary

This is a bad option, I don't see it being useful. It'll capture everything the MUD misspells, it'll capture what the players misspell in communication, and so on. I don't see this feature being any good :confused: 

I thought it was a way for server admins to tell Mudlet what words are legit for their game, such as fantasy names and so on - yes, really weird edge case I know...

SlySven commented:
> Explain to me again why this should be profile specific? If I have multiple profiles connected to a same game, they all need to learn their words seperately? Not sure how that's benefiting?!

Just because there is more than one profile there is no reason why they should related to the same MUD - or even be in the same language. I am sure Vadi, when he gets the chance to Mud doesn't confine himself just to English (American) Muds!

It **would** be possible (at least in a POSIX environment) to have a shell script (or other means) to "common" them *off-line* e.g. take the `profile.dic` files from several profiles, and, after chopping off the first line in each of them with `tail -n +2`, to `cat` them into one, `sort` them, `uniq -u` them to remove duplicates and then `wc -l` to count the number of lines so that it can be reinserted at the top of the first line of the composite.  It would not be a good idea to try and share the dictionaries between profiles with a (symbolic) link ***unless*** the profiles are ***never*** opened at the same time, otherwise the second profile to close will overwrite any changes made via the first profile.

vadi2 commented:
Well... if we think about it - what would happen if it was just Mudlet global? It'd allow a Russian word on an English MUD then, but that's not a big deal...

SlySven commented:
> This is a bad option, I don't see it being useful. It'll capture everything the MUD misspells, it'll capture what the players misspell in communication, and so on. I don't see this feature being any good :confused:.

It isn't going to work for everyone - which is why it is *optional* and *disabled* by default.  However it  can speed up an initial population of the dictionary as the user does not need to manually add each word themself - also it does capture all those odd Mud specific words permanently, unlike the current completion code that only looks back 500 lines and forgets everything between sessions.  The data *is* available to the lua sub-system which *can* modify it. For instance I have noticed in testing that it captures WoTMUD object/person location codes which are long, fixed length encoded strings that no-one apart from the MUD coders knows the encoding method of (to prevent manual faking) which some players can use to teleport (near) to the coded location - I would probably write a script to periodically purge them.

Also as I note - after a session a player can edit the `profile.dic` file manually with a text editor - and it is sorted alphabetically which makes that easier to do.

vadi2 commented:
This seems like a really obscure option a few people will use... and when they do use it, it'll create maintenance work for them too :confused:

vadi2 commented:
I'm personally not sold on it, but maybe others think it's an amazing idea, let's gather feedback from more users on this.

SlySven commented:
I was recently looking at HoloTrak - and that is very strict about you maintaining separation between alternate characters - and you can get penalised AFAICT for using names for other players/things that one character knows when another of your characters does not; in that case keeping specific spelling data confined to one profile will prevent you committing RP violations.

SlySven commented:
> I thought it was a way for server admins to tell Mudlet what words are legit for their game, such as fantasy names and so on - yes, really weird edge case I know...

That is feasible but would need to be done off-line - "download this file and append it to your `profile.dic`" file. If there *was* a demand for it we (I) could implement it ("Import a word list/dictionary into this profile")...

vadi2 commented:
Hm... I'm with @Kebap, I don't see why we shouldn't have it be per-application. The RP violation argument isn't strong since Mudlet doesn't do automatic replacements, in the end it's up to the user.

Kebap commented:
I don't feel too strongly about this "populate dictionary with game data" functionality, but tend to welcome it. Don't see it harm much, and may be helpful to the initiated. Not sure how many people will find and activate it though.

Then again, this PR is a nice example of maybe discussing the scope of changes before actually coding them, to prevent coding superflous stuff.

SlySven commented:
Can I propose a compromise in that I look to reworking the code to give each profile a three options radio button choice:

- [X] - No additional user editable dictionary
- [ ] - Per profile independent dictionary (`profile.dic`/`profile.aff`)
- [ ] - Common dictionary (`mudlet.dic`/`mudlet.aff`)

Personally I prefer to have a per profile but I am guess my inclinations are not everyones. :man_shrugging: 

:bulb: One thought that might make it easier to share words between independent profiles is to raise a Mudlet event within the profile every time a word is added or removed from that profile's dictionary whether done by the automatic capture, command line context menu or lua command.  An addition type event would also make it possible to filter the automatic added words in a finer manner than just on length and not having digits.

:thinking: As for the proposed highlighting I think you have to differentiate between where the "this word is okay" message comes from. Previously, flagging a word when it *isn't* in a system dictionary is satisfactory when there is only two states for that word - however now there is three states:
* the word is in the system dictionary (it could *also* be in the user dictionary but that does not alter the fact that the word is correctly spelled and *I think* flagging this is unhelpful)
* the word is not in the system dictionary - so either it is not a valid word or it is a game/Mud specific word - which leads to:
    * the word is in the user's dictionary - so it is probably valid
    * the word is not in the user's dictionary - so EITHER it is not valid OR it needs to be added to the dictionary

Logically though - it is NOT possible for the Mudlet application to distinguish between the EITHER and the OR case - therefore the only state that *can* be detected is the previous one where the word *IS* in the user's dictionary.  Which is how my code has it - Q.E.D.

vadi2 commented:
Can we not make simple spellchecking a super complicated feature... ?

How is it that every popular application out there gets by with a simple "Add to dictionary" option and not much more.

I'm not happy at all with this proposal now, it is getting more and more complicated to use; Mudlets strength is on being as _simple_ as possible, this is taking it completely the opposite direction by making spellchecking as _complicated_ as it can possibly be. We're now talking about raising events, highlighting the word in 2 different ways, we have a menu with spellchecking with two lists, and we even have a Lua API for spellchecking.

Like, let's take a step back and think about the roots of Mudlet and what got it to where it is today: people like it because (as much as we can make it be) it is **simple** and "just works".

vadi2 commented:
Let's analyse the situation: what's missing in Mudlet in regards to spellchecking is the ability to store your own words.

If we add a right-click option to store a custom word, we're done. That's it. Think of how simple spellchecking in a text processor or a web browser is. 

If that's not enough and there is interest for more, let's analyse this in detail before adding more features - or what would some people critisize in Mudlet as bloat - carefully.

Can we get away with one global dictionary, or do need to complement it with a per-profile dictionary as well? @SlySven do you feel strongly about it, or would it be ok to try gaming with a global dictionary first?

SlySven commented:
I feel strongly that one profile one dictionary is not unreasonable but I can see that it does not meet everyone's expectations.

I *am* willing to simplify things by removing the auto -populate the stuff from the MUD server output - it is, perhaps not as helpful as I would like because of the difficulty of filtering out what are the good words to add.

Instead, as IIRC `<Cntrl>`+left mouse click selects the whole line, I would in the future, consider a *separate* PR to provide a means to select only a single *word* from `TConsole` and add a context menu item to add, or if it is in the user's personal dictionary - to remove it from there. That would seem a more helpful UI in that the user can pick and add just the words they recognise that they might need to remember the spelling of in both the MUD output and in their command input.

vadi2 commented:
That sounds like a good change.

On the dictionaries: what if we simplify it by doing this: having a per-profile option to toggle use of application-wide (default) or profile-wide dictionary.

I think that would satisfy @Kebap's usecase (and the majority's too, I believe) while allowing per-profile dictionaries for those who want it.

SlySven commented:
Okay, I have polished this a bit further:
- I've taken out the grab words from the game server output
- I've added the three options (none/profile/shared) for the user dictionary
- I've found as many dictionary files as I can from my Debian and FreeBSD systems and added text identifications for them rather than just a short code (although that is reported along with a request to report it for ones that are not identified) - the tool tip for each one does identify the actual filePathName pair (.dic/.aff) that are being used which the user can use to identify them if needed.  I've revised the display to make better use of the space (no longer is it a single long list, now it also scrolls to the selected item when first shown).

The shared dictionary was a little harder to do than I thought as I felt it was necessary to do some read/write resource locking to prevent one profile from adding to the shared word list (actually a `QSet<QString>` as that is a better container than a `QStringList`) whilst another is trying to read it.

The lua functions check for the relevant functionality before trying to use them - i.e. they will nil + error message if the user dictionary is not enabled and the function was called in a manner that would use it.

![screenshot_20190218_005004](https://user-images.githubusercontent.com/6163092/52922347-94c9de00-3318-11e9-9576-048b7ae4664f.png)

I still haven't set spell checking to be turn on by default because Hunspell dictionaries are not available for every GUI language that we currently offer - and I do not yet know whether the MacOs case is handled - according to the code for that OS we only ever look in `QStringLiteral("%1/../Resources/").arg(QCoreApplication::applicationDirPath())` and it is not clear to me that we actually bundle anything there - if a MacOs tester can tell me that they do see dictionaries (perhaps our MacOs build scripts do put Hunspell files there or symlinks to the system's ones) then I will reconsider...

SlySven commented:
Just for comparison - this is what the currently development branch looks like on opening for me:
![screenshot_20190218_021458](https://user-images.githubusercontent.com/6163092/52924129-335b3c80-3323-11e9-8914-eee906148dba.png)

vadi2 commented:
We've had spellchecking on by default and nobody was complaining; let's still leave it on?

vadi2 commented:
This is a huge PR - I'm reviewing it, but it'll take some time

SlySven commented:
Just for comparison - this is what the currently development branch looks like on opening for me:
![screenshot_20190218_021458](https://user-images.githubusercontent.com/6163092/52924129-335b3c80-3323-11e9-8914-eee906148dba.png)

SlySven commented:
In hindsight I see that the default is for "English (United States)" and we do include a copy of the Hunspell files for THAT (even on macOs) so it would be okay to have it enabled as before - (there is a tool tip for the `QListWidget` when **no** dictionaries at all are found - but it *should* be quite hard to trigger that now...)

I will change that default back to being on but I'll leave that to avoid rocking the boat until you have digested what is already there.

vadi2 commented on src/TConsole.h:
```diff
@@ -298,8 +320,57 @@ public slots:
+
+    // Was public in Host class but made private there and cloned to here
+    // (for main TConsole) to prevent it being changed without going through the
+    // process to load in the the changed dictionary:
```
I don't understand this comment - why do the prevention?

SlySven replied:
Changing that value needs to do more than just changing that value - forcing the preference dialogue to change it via `Host::setSpellDic(const QString&)` ensures that the appropriate dictionary is loaded via a call to `TConsole::setSystemSpellDictionary(const QString&)` for the main `TConsole`.

vadi2 replied:
Ok - makes sense - could you revise it to be simpler to understand, ie "changing this value must also trigger that action"?

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -15592,3 +15597,227 @@ int TLuaInterpreter::disableClickthrough(lua_State* L)
+    Host& host = getHostFromLua(L);
+    if (!host.mEnableSpellCheck) {
+        lua_pushnil(L);
+        lua_pushstring(L, "spell-checking is not enabled in the preferences for this profile");
```
Is this an architectural limitation? Because otherwise I don't see why should the Lua API be off when the user has disabled spellchecking.

SlySven replied:
This seems like a least surprising option as far as I see it!

if they *have* switched off spellchecking they probably have not bothered (or cannot) set a reasonable dictionary for their locale (e.g. China) therefore it seems reasonable to switch it off here as well - and handle requests in reasonable run-time behaviour (i.e. with a nil and a tell them why message).

Also bear in mind that I will be reverting the change in that option so that it is enabled (the language will be ser for the "English (American)" which we bundle by default). So this will be an issue only for someone who has turned it off themselves anyway.

SlySven replied:
> Also bear in mind that I will be reverting the change in that option so that it is enabled (the language will be ser for the "English (American)" which we bundle by default)

Actually it seems as if it is not turned off even though I thought I had done so. Ah well maybe I had undone something already. :man_shrugging: :

SlySven replied:
It turns out that the control was unchecked so as to show that when there is not a profile loaded - it takes on the current setting of the `Host` instance when that is loaded in. And **that** does default to being enabled...

vadi2 replied:
I think of the API and Mudlets behaviour as separate - you should be able to use it while the preference is off. It's like we have a limitation where you can't use any of the map functions while the map is off.

SlySven replied:
Okay, that is now in place.

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -15592,3 +15597,227 @@ int TLuaInterpreter::disableClickthrough(lua_State* L)
+    int wordCount = wordList.size();
+    if (wordCount > 1) {
+        QCollator sorter;
+        sorter.setCaseSensitivity(Qt::CaseSensitive);
```
Why case-sensitive? Case-insensitive is more people-friendly

SlySven replied:
:thinking: Actually, I am not sure - and the Qt documentation does not indicate which is the default...

... okay will change to `Qt::CaseInsensitve`.

SlySven replied:
in next commit.

vadi2 commented on src/dlgProfilePreferences.cpp:
```diff
@@ -439,47 +440,159 @@ void dlgProfilePreferences::initWithHost(Host* pHost)
         }
+
+        // Reenable sorting now we have populated the widget:
+        dictList->setSortingEnabled(false);
```
```suggestion
        dictList->setSortingEnabled(true);
```

SlySven replied:
:man_facepalming: Yeah! - Will fix that.

SlySven replied:
In next commit.

vadi2 commented on src/mudlet.cpp:
```diff
@@ -147,6 +148,250 @@ void mudlet::loadLanguagesMap()
+    // https://www.unicode.org/cldr/charts/latest/supplemental/language_territory_information.html
+    // Initially populated from the dictionaries provided within the Debian
+    // GNU/Linux distribution:
+    //: In the translation source texts the language is the leading term, with, generatlly, the (primary) country(ies) in the brackets, with a trailing language disabiguation after a '-' Chinese is an exception!
```
`generally`

SlySven replied:
Yeah - well fix.

vadi2 commented on src/mudlet.cpp:
```diff
@@ -4306,3 +4559,501 @@ QPair<bool, QStringList> mudlet::getLines(Host* pHost, const QString& windowName
+    return Hunspell_create(affixPathFileName.toUtf8().constData(), dictionaryPathFileName.toUtf8().constData());
+}
+
+Hunhandle* mudlet::prepareSharedDictionary()
```
I don't understand what this and the previous function are trying to do. It seems they're trying to do too much at once. Could you split them up into more logical pieces?

SlySven replied:
Does the previous comment explain at all what is going on? If not I can explain in more detail.

SlySven replied:
Ah, given that there are common elements in all of:
* `mudlet::prepareProfileDictionary(...)`
* `mudlet::prepareSharedDictionary()`
* `mudlet::saveDictionary(...)`

I *have* been able to refactor them to call some common smaller methods, in the next commit.

SlySven commented on src/ui/profile_preferences.ui:
```diff
@@ -568,27 +568,93 @@
+                <bool>true</bool>
+               </property>
+               <property name="toolTip">
+                <string>&lt;p&gt;A user dictionary will not be used.&lt;p&gt;</string>
```
Well we didn't have it in the past and maybe some people like it like that... after all *there's nowt so queer as folk*!

SlySven replied:
... and even now I do not know that the user dictionary will work as expected for, say Arabic or RTL scripts. As it happens although spell-checking *is* enabled by default and set for "English (American)" the user dictionary is disabled by default to reflect how the previous versions of Mudlet are - I am sure you will want to draw attention to it in the release announcement... :mega:

vadi2 replied:
If it doesn't work people can turn spellchecking off or we'll fix it. I don't think this option is a super-useful one to have - and we already have an "options overload" with so many knobs and fobs, let's stick to adding ones we're super sure are needed?

SlySven replied:
It doesn't seem that sensible to *force* the user to have a personal dictionary (to hold their own curated lists of words for things like persons' names that are never going to fit into a national one) if they don't want it. Don't forget that it will take up a couple of lines on the context menu even if it cannot contribute because it has "no suggestions".  It really doesn't seem that unreasonable to have a three way choice for: shared between all profiles that want it / just for this per profile / not at all to shut it off completely. The latter will make things be how they are before I came up with this plan...

vadi2 replied:
It's only how all applications work today - Firefox, LibreOffice and so on. You will not find a 'personal dictionary' option in them!

SlySven commented on src/TBuffer.cpp:
```diff
@@ -5093,3 +5094,20 @@ QString TBuffer::processSupportsRequest(const QString& elements)
 }
+
+// Count the graphemes in a QString - returning it's length in terms of those:
+unsigned int TBuffer::graphemeLength(const QString& text)
```
:thinking: Humm, in this case I guess it could be okay to have ~`singed`~  `signed int`s.

However in other places in the `TBuffer` class it is a real PITA that the formatting attributes are in a `std::deque<std::deque<TChar>>` type which use the **`unsigned`** `std::size_t` (directly from the C `size_t` type) for indexing which is by definition always the right size to use for indexing into arrays whereas the textual content is based upon `QList<QList<QChar>>` or rather `QStringList` which uses `int`s for indexing. Making the code seem type safe (and getting rid of all the clang code warnings about implicit casts between signed and unsigned integer types) with explicit `static_cast<Ta>(value of type Tb)` is not my idea of fun...

SlySven replied:
In next commit.

As it happens, with the removal of the code to add words directly from the game server's output this method is not currently used.  I expect however to need it in the very near future (along with a `graphemeMid(int start, int length = -1)`) to better handle the text selection/highlighting issues that we have had for a long time now and which I promised to get onto, was it, *last* week. :worried:

SlySven commented on src/TBuffer.cpp:
```diff
@@ -5093,3 +5094,20 @@ QString TBuffer::processSupportsRequest(const QString& elements)
+    unsigned int count = 0;
+    while (pos > 0) {
+        ++count;
+        pos = graphemeFinder.toNextBoundary();
```
Well `lengthInGraphemes` is a *longer* name than `graphemeLength` and in **English** (to the best of my knowledge) the current term does (can be?) read as `grapheme`(units thereof, i.e. measured in that *unit*)`Length` - but I can switch to `lengthInGraphemes` (or `sizeInGraphemes`) if you insist. ***Do*** you insist? :grin:

vadi2 replied:
I think that's a much more intuitive name, yes!

SlySven replied:
OTOH it looks irregular compared to the `graphemeMid(...)` method I promulgated above...

SlySven replied:
Renamed to `lengthInGraphemes` in next commit - as it happens I don't need it for this PR any more but I want to have it around in the near future...

SlySven commented on src/TCommandLine.h:
```diff
@@ -91,11 +89,11 @@ public slots:
-    int mHunspellSuggestionNumber;
-    char** mpHunspellSuggestionList;
+    QString mSpellCheckedWord;
+    int mHunspellSystemDictionarySuggestionsCount;
```
Fair comment - and I *think*... yes, we can.  There will be another, similar, member which can become `mUserDictionarySuggestionsCount` I guess.

SlySven commented on src/TConsole.cpp:
```diff
@@ -550,7 +552,7 @@ TConsole::TConsole(Host* pH, ConsoleType type, QWidget* parent)
 
     changeColors();
-    if (mType & (MainConsole|Buffer)) {
+    if (mType == MainConsole) {
```
I am picking away at them as I find them. To me this is an *obvious* case where previous code did not consider the `Buffer` case and I think it will be safe to drop it from **here**... :boom:

SlySven commented on src/TConsole.cpp:
```diff
@@ -561,6 +563,29 @@ TConsole::TConsole(Host* pH, ConsoleType type, QWidget* parent)
+        // Load up the spelling dictionary for the profile - needs to handle the
+        // absence of files for the first run in a new profile or from an older
+        // Mudlet version:
+        setProfileSpellDictionary();
```
Yeah, it was a bit of a note to myself as well - and is a bit out-of-date anyhow - the `.aff` file is created anew each time now...

... but I *should* put the comment in the *right* place. :eyes:

SlySven replied:
in the next commit.

SlySven commented on src/TCommandLine.cpp:
```diff
@@ -42,12 +42,11 @@ TCommandLine::TCommandLine(Host* pHost, TConsole* pConsole, QWidget* parent)
+, mHunspellSystemDictionarySuggestionsCount()
+, mHunspellProfileDictionarySuggestionsCount()
+, mpHunspellSystemSuggestionsList()
+, mpHunspellProfileSuggestionsList()
```
Swings-and-roundabouts IMHO - it made it easier to track the members I was working on with them all having that in them...

... but as I think things are largely sorted out :pray: - I can now take it out.

vadi2 replied:
What does that expression mean?

And, cool!

SlySven replied:
:information_source: Short for "What you gain on the swings you lose on the roundabouts" - https://english.stackexchange.com/q/216204 .

vadi2 replied:
Aha, thanks :)

SlySven replied:
Those four methods renamed without the Hunspell in next commit.

SlySven commented on src/TCommandLine.cpp:
```diff
@@ -524,20 +516,40 @@ void TCommandLine::adjustHeight()
+                // use a different underline, on many systems the spell-check underline is
+                // a wavy line but on macOs it is a dotted line - so use dash underline
+                f.setUnderlineStyle(QTextCharFormat::DashUnderline);
+                f.setUnderlineColor(Qt::cyan);
```
Well ***I*** want to know if it is a word I have okayed (i.e. is something that has been added from a MUD or personal whim) or it has come from the main dictionary - if it is the **former** then I can change it - and I can tell that without having to right click and wait (there is a little delay remember to do the additional `Huspell_suggest(...)` calls.  I would hesitate to say I can make it optional but that is quite *do-abIe* if you cannot agree to this.

The choice of making it a dashed underline cyan was so that it was a *reassuring* greenish colour and clearly not the style of a spelling error (but not so green that it would be a problem for red-green colour bindness - although a bright blue would be better that would not contrast with the selection blue background; the different underline style was also chosen as a fallback if the colour was not enough).

vadi2 replied:
What do others on team think?

SlySven commented on src/TCommandLine.cpp:
```diff
@@ -566,56 +578,181 @@ void TCommandLine::slot_popupMenu()
+                action_removeWord->setEnabled(false);
+//                }
+                if (mudlet::self()->mUsingMudletDictionaries) {
+                    action_dictionarySeparatorLine = new QAction(tr("▼Mudlet▼ │ dictionary suggestions │ ▲User▲",
```
As before - some users (like myself) will want to know if it is a word from the MUD or added.

Perhaps it is less of an issue when such a word has been added by hand, but, for instance I would anticipate having triggers to add the names of all the players I identify in MUDs I play to that profile's dictionary so I always have them stored even if I haven't encountered them this session (so they are not in the tab completion list).  As there is considerable scope for automation via scripts in this (I think) - I do not think it is unreasonable to keep the two sets of words distinct - in fact - they have to be because they come from two completely unrelated sets of Hunspell files/dictionaries and merging them would actually take more code I believe.  Of course I could remove the decoration permanently (as happens if the user one is disabled) so there would seem to be just one list of suggestions - it is just that with the user one enabled the list could be up to twice the size as each is subject to a compiled in Hunspell limit.

I wonder if the @Mudlet/ui-review team have any input?

SlySven commented on src/TConsole.cpp:
```diff
@@ -2594,3 +2619,206 @@ void TConsole::slot_reloadMap(QList<QString> profilesList)
+    return result;
+}
+
+QPair<bool, QString> TConsole::removeSingleWordFromSet(const QString& word)
```
It ***is*** a `QSet<QString>` - for speed of checking for inclusion/so that there is no need to avoid the duplication that a `QStringList` (or `QList<QString>`) would require - so I felt calling it `removeSingleWordFromList` was *less* intuitive/consistent.

The name also reflected that there was a (now banished) method to add all the words in a line (that were not in the main dictionary AND were above a certain number of graphemes long AND {optionally} did not contain numbers) of text to the same container.

OTOH It does make sense now to remove the `single` from the `add...` and `remove...` methods.

SlySven replied:
Have removed the single in next commit.

SlySven commented on src/TConsole.cpp:
```diff
@@ -2594,3 +2619,206 @@ void TConsole::slot_reloadMap(QList<QString> profilesList)
+QPair<bool, QString> TConsole::removeSingleWordFromSet(const QString& word)
+{
+    QString errMsg = QStringLiteral("the word \"%1\" does not seem to be in the user dictionary");
+    QPair<bool, QString> result = qMakePair(false, QString());
```
:thinking: Actually it might - it is an initialisation and *one* of the pair will be modified before it is passed out via Return-Value-Optimisation...

SlySven replied:
It does and has been done.

SlySven commented on src/Host.h:
```diff
@@ -121,6 +121,8 @@ class Host : public QObject
     void               setSpellDic(const QString&);
     const QString&     getSpellDic() { QMutexLocker locker(& mLock); return mSpellDic; }
+    void               setUserDictionaryOptions(const bool useDictionary, const bool useShared);
+    void               getUserDictionaryOptions(bool& useDictionary, bool& useShared) { QMutexLocker locker(& mLock); useDictionary = mEnableUserDictionary; useShared = mUseSharedDictionary; }
```
I need to get both values in some cases and doing it via non `const` references seemed a valid (and more atomic - only one locking action) way to do it. The getter needs the lock to prevent it being accessed at the same time as the setter is changing it IMVHO.  Annoyingly we have not overhauled the `Host` class to use a `QReadWriteLock` instead to allow multiple *readers* to get details without stepping on each others toes. Sadly some of the recent code added into the Mudlet project has not even bothered to pay attention to `Host::mLock` so I expect more *nasal daemons* to emerge in the future...

N.B. remember that the spell-suggestions code is triggered by context menu (or at least user right mouse) click code so is asynchronous (*I think*) and is the sort of thing that might come to bite us in the future if we do not care to think about thread safety...

SlySven commented on src/TBuffer.cpp:
```diff
@@ -5093,3 +5094,20 @@ QString TBuffer::processSupportsRequest(const QString& elements)
     return result.join(QLatin1String(" "));
 }
+
+// Count the graphemes in a QString - returning it's length in terms of those:
```
Ah, that is when it's not to be apostrophed :laughing:

:+1: on the grammar catch.

SlySven replied:
In next commit.

SlySven commented on src/TCommandLine.cpp:
```diff
@@ -566,56 +578,181 @@ void TCommandLine::slot_popupMenu()
-            if (mHunspellSuggestionNumber) {
-                for (int i = 0; i < mHunspellSuggestionNumber; ++i) {
-                    auto pA = new QAction(mpHunspellCodec->toUnicode(sl[i]));
+            mHunspellSystemDictionarySuggestionsCount = Hunspell_suggest(handle_system, &mpHunspellSystemSuggestionsList, encodedText.constData());
```
Ah, yeah a hang-over from when the value was stored as `mHunspellSuggestionNumber` - which it actually wasn't...

... comment to be excised :hocho: :smiling_imp:

SlySven replied:
In next commit.

SlySven replied:
Done.

SlySven commented on src/TConsole.cpp:
```diff
@@ -2594,3 +2619,206 @@ void TConsole::slot_reloadMap(QList<QString> profilesList)
+
+void TConsole::setSystemSpellDictionary(const QString& newDict)
+{
+    if (!newDict.isEmpty() && mSpellDic != newDict) {
```
:bulb: Yeah, I think you're right.

... will do.

SlySven replied:
In next commit.

SlySven commented on src/TConsole.cpp:
```diff
@@ -2594,3 +2619,206 @@ void TConsole::slot_reloadMap(QList<QString> profilesList)
+        if (QFile::exists(QStringLiteral("/usr/share/hunspell/%1.aff").arg(currentDictionary))) {
+            path = QLatin1String("/usr/share/hunspell/");
+            mudlet::self()->mUsingMudletDictionaries = false;
+        } else if (QFile::exists(QStringLiteral("%1/../../src/%2.aff").arg(QCoreApplication::applicationDirPath(), currentDictionary))) {
```
I think it is an OS implementation detail as to where *they* keep their dictionaries and for *Mudlet* it is depending on what sort of build it is - if an installer version we will want to use the ones we bundled (if any, and there not being any already on the system we can use which are more likely to be to the users own requirements).

In reality we are looking around to find the best available ones and they are listed in descending order of likely usefulness (IMHO) given all the different platforms we are working with. Once we find something then we are happy (and that is why I indicate the locations in the individual tooltips - just in case the end user *does* have multiple sets of dictionaries around, like say, if they are working on the source code and want to know what is going on...)

vadi2 replied:
Wouldn't it be better to collect the most amount of dictionaries possible?

SlySven commented on src/TConsole.cpp:
```diff
@@ -2594,3 +2619,206 @@ void TConsole::slot_reloadMap(QList<QString> profilesList)
+const QSet<QString>& TConsole::getWordSet() const
+{
+    if (!mEnableUserDictionary) {
+        QSet<QString>();
```
:beetle: Ooops, missing a `return` there to handle the case when there isn't a user dictionary in use.

SlySven replied:
In next commit.

SlySven commented on src/mudlet.cpp:
```diff
@@ -147,6 +148,250 @@ void mudlet::loadLanguagesMap()
+    // Initially populated from the dictionaries provided within the Debian
+    // GNU/Linux distribution:
+    //: In the translation source texts the language is the leading term, with, generatlly, the (primary) country(ies) in the brackets, with a trailing language disabiguation after a '-' Chinese is an exception!
+    mDictionaryLanguageCodeMap = {{QStringLiteral("af"), tr("Afrikaans")},
```
In the future it *might* be possible to merge this with the GUI translation map but only if we consider allowing run-time GUI language changes and some additional code to mangle the output into a form suitable for both cases.

In the meantime I think I may have put this in the wrong place - will need to check the program flow through this method...

SlySven replied:
... okay, it is alright where it is I think.

SlySven commented on src/mudlet.cpp:
```diff
@@ -4306,3 +4559,501 @@ QPair<bool, QStringList> mudlet::getLines(Host* pHost, const QString& windowName
     }
 }
+
+Hunhandle* mudlet::prepareProfileDictionary(const QString& hostName, QSet<QString>& wordSet)
```
Not IMHO -  the dictionary file `.dic` (list of words as we need it) is likely to already exist if it has been used before - however this takes that file which ***can** have been **manually** edited by the user* - and rechecks the word count on the first line and recalculates the grapheme usages within so that the `TRY ` line in the `.aff` file can be revised to match the current contents of the `.dic` file - which is needed for `Hunspell_suggest` to work and to work in an optimum manner.

As it happens the existing `.aff` file is replaced here (and in the `saveDictionary` call).

Anyhow - this is about preparing the files for use by the Hunspell library - and it only **creates** the `.dic` on first usage so *I* think the current name is more appropriate.

SlySven commented on src/mudlet.h:
```diff
@@ -361,12 +366,26 @@ class mudlet : public QMainWindow, public Ui::main_window
     QPair<bool, QStringList> getLines(Host* pHost, const QString& windowName, const int lineFrom, const int lineTo);
+    void setEnableFullScreenMode(const bool);
+
+    // Both of these revises the contents of the .aff file: the first will
```
Comment needs editing - as there are now *three* methods involved and it is the third, `saveDictionary`, that replaces the contents of `.dic` with those of the supplied `QSet<QString>` (and the contents of `.aff` with details derived from that replacement).

SlySven replied:
... in next commit.

SlySven commented on src/TConsole.cpp:
```diff
@@ -2594,3 +2619,206 @@ void TConsole::slot_reloadMap(QList<QString> profilesList)
+        QString path;
+        const QString& currentDictionary = mpHost->getSpellDic();
+#if defined(Q_OS_MACOS)
+        path = QStringLiteral("%1/../Resources/").arg(QCoreApplication::applicationDirPath());
```
Okay, in the next commit use `mudlet::getMudletPath`  with a first argument of `hunspellDictionaryPath` and an extra argument of the dictionary affix file to look for - e.g. "en_US" for the default case.  The revision I am just coding will also set `(bool) mudlet::mUsingMudletDictionaries` depending on where it finds the files as current done in this PR.

SlySven commented:
Need to make a change to the `TConsole` class (rename the `QString` `TConsole::profile_name` to be `TConsole::mProfileName`) so that it is compatible with the other PR.

vadi2 commented:
@keneanung or @Kebap would you have time to review this & offer opinion?

vadi2 commented on src/ui/profile_preferences.ui:
```diff
@@ -525,30 +525,93 @@
+          <item>
+           <widget class="QCheckBox" name="checkBox_spellCheck">
+            <property name="toolTip">
+             <string>&lt;p&gt;This option controls spell-checking on the command line in the main console for this profile.&lt;/p&gt;&lt;p&gt;&lt;i&gt;The dictionary selected below will be available to the lua sub-system whether this is checked or not.&lt;/i&gt;&lt;/p&gt;</string>
```
> The dictionary selected below will be available to the lua sub-system whether this is checked or not.

This is a good tidbit of information but I don't think this is the right place for it, documentation for the Lua API would be better.

SlySven replied:
It does help to clarify the action of the control - so why is it *not* a right place for it? :confused: Indeed if the user only wants the dictionary for the lua subsystem (which is a bit of an odd case I admit) then this will reassure them as to its effects.

vadi2 replied:
I see your point in being helpful to explain what the control does - but most people aren't coders who'll understand what does "lua sub-system" mean, right?

It's not the right place for it because if you want information on the Lua functionality, you look in the documentation, not the tooltips. It's the same reason we don't talk about the Lua API for the mapper on any of the mapper-related dialogs or menus.

SlySven replied:
Okay - I'll take out the *second* paragraph.

vadi2 commented:
So following https://github.com/Mudlet/Mudlet/pull/2358#issuecomment-463141374 - I've been talking to people out in the real world about this custom cyan highlighting and the idea does not have much support. Is it something you'd still like @SlySven?

SlySven commented:
I still want it to know that it is a word that has been stored in my personal dictionary (so it could be a deliberate *mis-spelling* for effect perhaps) rather than one from the system. In the light of your previous ~~rant~~indication that you'd prefer less rather than more knobs why don't we give it a spin in the field and see if it is *that* objectionable before we have to add a control for it.

*\<aside>I don't suppose it is helpful to point out that* one *way to make the cyan dashed underline go away is to turn **off** the personal dictionary?\</aside>* :wink:

vadi2 commented:
Sure - let's see what the testers think of it!

SlySven commented:
Can we just establish which of those are being introduced by this PR and which were there already. I am not sure that changing the dictionary caused a recheck in the past either - so this is merely highlighting an existing defect rather than introducing a new one.

Default *personal* dictionary is `None` so as to not automatically introduce new functionality (including those cyan underlines :grinning:) without the user asking for it - then they can decide whether they want to have a common dictionary or per profile ones - after all it seems even ***we*** have different preferences in this so why would the users not also have them?

vadi2 commented:
Do you have any examples of applications that have personal dictionaries as an option that is off by default?

vadi2 commented:
The reason I ask is - default options **really** matter, both in software design and outside it, and we have to really think on what good defaults are.

SlySven commented:
You do have the option to turn off the *standard* but not the *ignoreAllList* (probably for technical reasons so that `Ignore All` works)  in **LibreOffice**. The `en-US` and `en-GB` are language specific user dictionaries that I have chosen to install and the `standard` is an any-language one which I think is always present but they are ***all*** for the user's words that are not in the standard (system) or bundled language (not sure which it is on FreeBSD) dictionaries.  Given that this is a word-processing example I can draw parallels between them having individual language specify AND a *shared* common user dictionaries (the latter would probably be used for person names for instance that do not change between langauges) to us having a per-profle/MUD and a common dictionary option - whilst I *could* code for Mudlet to provide BOTH at the same time I think that will be more effort than virtually all users would care for...

![screenshot_20190306_175501](https://user-images.githubusercontent.com/6163092/53902396-ff795a00-4038-11e9-98a1-eb3a18af2c61.png)

vadi2 commented:
I could be misinterpreting, but from what I can see LibreOffice does not allow you to turn off custom user dictionaries? You can just enable/disable system ones.

SlySven commented:
Those are all the user ones - they are independent of the Hunspell system ones - which - being in the `/usr/` part of the filesystem are shared read-only. LibreOffice gives you the option of a (actually you could have multiple ones) non-language specific one (the `[all]` one) or ones that are used depending on which language the section of text is marked as being in. (the `[English (UK)]` and `[English (USA)]` ones).

If it is *really* going to be a sticking point (like if you are Donald Tusk and I am playing Theresa May wanting to change the *Backstop* arrangement :wink:) then I guess I will concede the point and remove the `None` option but I don't want to do it... :slightly_frowning_face:

vadi2 commented:
Let's leave None in, okay, but set the default to Shared? That would mimic LibreOffice's default.

SlySven commented:
Ho, humm, I am getting odd problems now - I had managed to hide the "None" option but now I am getting consistent `SIGILL`s at an odd point.

Oh, *finally* got some working and meaningful analyser output that seems relevant and I've got a pair of "*returning reference to a temporary object*" and a "*control reaches end of a non-void function*" issues to deal with - ah - that is so stupid of me! :man_facepalming:

SlySven commented:
- *changing dictionaries and hitting Save does not re-check existing words in the input line* - **DONE**
- *default user dictionary is none - should be shared* - I disagree - the safer/simpler code path is the per profile one because it does not involve extra code to prevent one profile reading the shared dictionary whilst another one is editing it - **SORT OF DONE** (the "None" option is hidden and not currently accessible)
- *adding a word to the user dictionary does not remove red highlight* - **DONE**
- *removing word from user dictionary does not remove cyan highlight* - **DONE**
- *spelling suggestions are still shown for valid words (valid either according to system or user dictionary)* - **not done** - suggestions need not be just for when there is a mistake - they can also help with *homophones* words that sound the same but have different spellings: the main dictionaries will usually include coverage for those as well.

The add/remove word things are not 100% - not quite sure of the mechanism for what is happening but it is better IMO than it was...

Kebap commented:
Microsoft Word uses multiple colors for different underlines:

Red = Spelling error
Green = Grammatical error
Blue = Spelling correct, but wrong context

![grafik](https://user-images.githubusercontent.com/117238/53953518-d22dba00-40d3-11e9-8f95-c8812d68ceb8.png)

They offer no user-defined dictionaries.

vadi2 commented:
(and all are errors, correct words are not underlined)

SlySven commented:
As I imagine it the main purpose of a user dictionary is so that words specific to a MUD can be reproduced (:bulb: so the option to use a context menu action to grab selected words from at least the `main` `TConsole` display will be a worthwhile future enhancement :grin: ) - that being the case the sharing of words between profiles is only going to be useful for MUDs where the user uses multiple profiles for different characters - in the more general case where the user plays different MUDs (or is a MUD client developer maintaining a number of MUD server accounts) or they use the same account for multiple characters and make that choice as part of a more complex login process than we (currently) support. It is not unreasonable IMVHO to keep the dictionaries for each profile as separate as all the other bits of data for that profile is.

The mechanism for sharing a dictionary between profiles is a bit more complex (I've put in a `QReadWriteLock` to prevent awkwardness with one profile changing the dictionary whilst others want to read it) so I would recommand the independent case simply because it is less likely to have any issues... :crossed_fingers: 

As things stand it may be feasible to allow the users to specify shared dictionaries if they need them for multiple clients instead of the hard-coded `mudlet.`(`dic`|`aff`) pair in the base of the mudlet data directories. That is something that can be safely deferred for the moment I believe.

vadi2 commented:
Could someone pop this into the testers channel for feedback?
